compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/draganddrop/DragAndDropNode.kt

```diff
@@ -93,43 +97,80 @@ fun DragAndDropModifierNode(): DragAndDropModifierNode = DragAndDropNode { null
-    else null
+): DragAndDropModifierNode =
+    DragAndDropNode(
+        onDropTargetValidate = { event -> if (shouldStartDragAndDrop(event)) target else null }
+    )
+
+/** A scope that allows starting a drag and drop session. */
+internal interface DragAndDropStartTransferScope {
+    /**
+     * Initiates a drag-and-drop operation for transferring data.
+     *
+     * @param transferData the data to be transferred after successful completion of the drag and
+     *   drop gesture.
+     * @param decorationSize the size of the drag decoration to be drawn.
+     * @param drawDragDecoration provides the visual representation of the item dragged during the
+     *   drag and drop gesture.
+     * @return true if the method completes successfully, or false if it fails anywhere. Returning
+     *   false means the system was unable to do a drag because of another ongoing operation or some
+     *   other reasons.
+     */
+    fun startDragAndDropTransfer(
+        transferData: DragAndDropTransferData,
+        decorationSize: Size,
+        drawDragDecoration: DrawScope.() -> Unit,
+    ): Boolean
 }
 
 /**
- * Core implementation of drag and drop. This [Modifier.Node] implements tree traversal for
- * drag and drop, as well as hit testing and propagation of events for drag or drop gestures.
+ * Core implementation of drag and drop. This [Modifier.Node] implements tree traversal for drag and
+ * drop, as well as hit testing and propagation of events for drag or drop gestures.
  *
  * It uses the [DragAndDropEvent] as a representation of a single mutable drag and drop session.
  *
- * The implementation implicitly maintains a sorted tree of nodes where the order of traversal
- * is determined by the proximity to the last event. That is, after finding a receiving node,
- * the next event will follow the same path the previous event did unless a fork is found and
- * another node should receive the event.
+ * The implementation implicitly maintains a sorted tree of nodes where the order of traversal is
+ * determined by the proximity to the last event. That is, after finding a receiving node, the next
+ * event will follow the same path the previous event did unless a fork is found and another node
+ * should receive the event.
+ *
+ * This optimizes traversal for the common case of move events where the event remains within a
+ * single node, or moves to a sibling of the node.
  *
- * This optimizes traversal for the common case of move events where the event remains within
- * a single node, or moves to a sibling of the node.
+ * This intended to be used directly only by [DragAndDropManager].
  */
 internal class DragAndDropNode(
-    private val onDragAndDropStart: (event: DragAndDropEvent) -> DragAndDropTarget?
-) : Modifier.Node(),
+    private var onStartTransfer: (DragAndDropStartTransferScope.(Offset) -> Unit)? = null,
+    private val onDropTargetValidate: ((DragAndDropEvent) -> DragAndDropTarget?)? = null,
```

Comment:
Can you document behavior of this lambda?

Reply:
Done